project-koku/koku
# Cost 2606

Fixes [COST-2606](https://issues.redhat.com/browse/COST-2606)

Changes proposed in this PR:
* Fix an error in an endpoint when utilizing filter[limit] and filter[offset] with desc delta
* Also added some validation to not use filter[limit]/filter[offset] with a group_by parameter also being added, except for on instance-types reports

Testing Instructions:
1. Pull branch
2. create and load test customer data
3. Hit the following endpoint should give a 200 response with the following data
http://127.0.0.1:8000/api/cost-management/v1/reports/aws/costs/?order_by[delta]=desc&delta=cost
   ```
    "meta": {
        "count": 10,
        "delta": {
            "value": -262.691891766,
            "percent": -8.786270270385916
        },
        "cost_type": "unblended_cost",
        "compute_count": false,
        "check_tags": false,
        "order_by": {
            "delta": "desc"
        },
        "filter": {
            "time_scope_value": "-10",
            "time_scope_units": "day",
            "resolution": "daily"
        },
        "group_by": {},
        "total": {
            "infrastructure": {
                "raw": {
                    "value": 2521.825379341,
                    "units": "USD"
                },
                "markup": {
                    "value": 205.282537936,
                    "units": "USD"
                },
                "usage": {
                    "value": 0.0,
                    "units": "USD"
                },
                "total": {
                    "value": 2727.107917277,
                    "units": "USD"
                }
            },
            "supplementary": {
                "raw": {
                    "value": 0.0,
                    "units": "USD"
                },
                "markup": {
                    "value": 0.0,
                    "units": "USD"
                },
                "usage": {
                    "value": 0.0,
                    "units": "USD"
                },
                "total": {
                    "value": 0.0,
                    "units": "USD"
                }
            },
            "cost": {
                "raw": {
                    "value": 2521.825379341,
                    "units": "USD"
                },
                "markup": {
                    "value": 205.282537936,
                    "units": "USD"
                },
                "usage": {
                    "value": 0.0,
                    "units": "USD"
                },
                "total": {
                    "value": 2727.107917277,
                    "units": "USD"
                }
            }
        }
    },
   ```
4. Hitting the following endpoint should now cause an error message since a group_by parameter is not being used with filter[limit] 

http://127.0.0.1:8000/api/cost-management/v1/reports/aws/costs/?order_by[delta]=desc&delta=cost&filter[limit]=10
```
{
    "error": [
        "filter[limit] and filter[offset] requires a valid group_by param."
    ]
}
```
5. The following endpoint should give a 200 response, instance-types automatically group_by instance type thus the filter[limit] should be allowed and executed 
http://127.0.0.1:8000/api/cost-management/v1/reports/aws/instance-types/?order_by[delta]=desc&delta=usage&filter[limit]=10
```
"meta": {
        "count": 10,
        "others": 0,
        "delta": {
            "value": -131.0,
            "percent": -7.7976190476190474
        },
        "cost_type": "unblended_cost",
        "compute_count": false,
        "check_tags": false,
        "filter": {
            "limit": 10,
            "time_scope_value": "-10",
            "time_scope_units": "day",
            "resolution": "daily"
        },
        "order_by": {
            "delta": "desc"
        },
        "group_by": {},
        "total": {
            "usage": {
                "value": 1549.0,
                "units": "Hrs"
            },
            "infrastructure": {
                "raw": {
                    "value": 767.944,
                    "units": "USD"
                },
                "markup": {
                    "value": 29.8944,
                    "units": "USD"
                },
                "usage": {
                    "value": 0.0,
                    "units": "USD"
                },
                "total": {
                    "value": 797.8384,
                    "units": "USD"
                }
            },
            "supplementary": {
                "raw": {
                    "value": 0.0,
                    "units": "USD"
                },
                "markup": {
                    "value": 0.0,
                    "units": "USD"
                },
                "usage": {
                    "value": 0.0,
                    "units": "USD"
                },
                "total": {
                    "value": 0.0,
                    "units": "USD"
                }
            },
            "cost": {
                "raw": {
                    "value": 767.944,
                    "units": "USD"
                },
                "markup": {
                    "value": 29.8944,
                    "units": "USD"
                },
                "usage": {
                    "value": 0.0,
                    "units": "USD"
                },
                "total": {
                    "value": 797.8384,
                    "units": "USD"
                }
            }
        }
    },...

```

---
Additional Context:
